cou/apps/app.py

```diff
@@ -293,6 +375,76 @@ def current_os_release(self) -> OpenStackRelease:
             f"{os_versions}. This is not currently handled."
         )
 
+    @property
+    def is_from_charm_store(self) -> bool:
+        """Check if application comes from charm store.
+
+        :return: True if comes, False otherwise.
+        :rtype: bool
+        """
+        return self.charm_origin == "cs"
+
+    @property
+    def is_subordinate(self) -> bool:
+        """Check if application is subordinate.
+
+        :return: True if subordinate, False otherwise.
+        :rtype: bool
+        """
+        return bool(self.status.subordinate_to)
+
+    @property
+    def is_os_channel_based(self) -> bool:
+        """Check if application is channel based.
+
+        :return: True if does have origin setting, False otherwise.
+        :rtype: bool
+        """
+        return (
+            not self.origin_setting
+            or self.is_versionless
+            or self.is_subordinate
+            or self.charm == "gnocchi"
+        )
+
+    def is_valid_track(self, charm_channel: str) -> bool:
+        """Check if the channel track is valid.
+
+        :param charm_channel: Charm channel. E.g: ussuri/stable
+        :type charm_channel: str
+        :return: True if valid, False otherwise.
+        :rtype: bool
+        """
+        if self.is_from_charm_store:
+            return True
+        try:
+            OpenStackRelease(self._get_track_from_channel(charm_channel))
+            return True
+        except ValueError:
+            return False
+
+    @staticmethod
+    def _get_track_from_channel(charm_channel: str) -> str:
```

Comment:
`CephMonApplication` also needs to get the track from channel for `_get_change_require_osd_release_plan` method. This is probably an indicator that we need to split the PR into multiple more-focused ones.

Reply:
Completely agree on this.